storybook/pagesmodel.cpp

```diff
@@ -1,45 +1,150 @@
 #include "pagesmodel.h"
 
-#include <QRegularExpression>
 #include <QDir>
+#include <QFileSystemWatcher>
+#include <QRegularExpression>
 
 namespace {
 const auto categoryUncategorized QStringLiteral("Uncategorized");
 }
 
 PagesModel::PagesModel(const QString &path, QObject *parent)
-    : QAbstractListModel{parent}
+    : QAbstractListModel{parent}, m_path{path},
+      fsWatcher(new QFileSystemWatcher(this))
 {
-    QDir dir(path);
-    dir.setFilter(QDir::Files);
+    m_items = load();
+    readMetadata(m_items);
+
+    fsWatcher->addPath(path);
+
+    connect(fsWatcher, &QFileSystemWatcher::directoryChanged,
+            this, &PagesModel::reload);
+}
 
+QList<PagesModelItem> PagesModel::load() {
     static QRegularExpression fileNameRegex(
                 QRegularExpression::anchoredPattern("(.*)Page\\.qml"));
-    static QRegularExpression categoryRegex(
-                "^//(\\s)*category:(.+)$", QRegularExpression::MultilineOption);
+
+    QDir dir(m_path);
+    dir.setFilter(QDir::Files);
 
     const QFileInfoList files = dir.entryInfoList();
+    QList<PagesModelItem> items;
 
-    std::for_each(files.begin(), files.end(), [this] (auto &fileInfo) {
+    std::for_each(files.begin(), files.end(), [this, &items] (auto &fileInfo) {
         QString fileName = fileInfo.fileName();
         QRegularExpressionMatch fileNameMatch = fileNameRegex.match(fileName);
 
         if (!fileNameMatch.hasMatch())
             return;
 
-        QFile file(fileInfo.filePath());
-        file.open(QIODevice::ReadOnly);
-        QByteArray content = file.readAll();
+        PagesModelItem item;
+        item.path = fileInfo.filePath();
+        item.title = fileNameMatch.captured(1);
+        item.lastModified = fileInfo.lastModified();
 
-        QRegularExpressionMatch categoryMatch = categoryRegex.match(content);
-        QString category = categoryMatch.hasMatch()
-                ? categoryMatch.captured(2).trimmed() : categoryUncategorized;
+        items << item;
+    });
+
+    return items;
+}
+
+void PagesModel::readMetadata(PagesModelItem& item) {
+    static QRegularExpression categoryRegex(
+                "^//(\\s)*category:(.+)$", QRegularExpression::MultilineOption);
+
+    QFile file(item.path);
+    file.open(QIODevice::ReadOnly);
```

Comment:
You should probably test here if the `open()` was successful; warn and do an early return

Reply:
There is no high need to check that. In worst case readAll will return empty array and it will be treated as "uncategorized".